[feature] Fix components delete in components list [OSF-8914]

fabmiz opened the pull request:
## Purpose

Allow deleting components with children from components list (project overview page).

Currently, trying to delete a component with children via the quickActionsMenu fails
because it was written before [OSF-8516](https://openscience.atlassian.net/browse/OSF-8516)

## Changes

- Add `NodesDeleteManager` view model to support both components and nodes deletion in both
project settings and components list (project overview page).
- Handle deleting node with/without NodesDeleteManager 
- Add ko `QuickDeleteViewModel` to allow customizing deleting node with no children
- Make `delete` button clickable only when user enters the right name.
- Make guide text red when user types wrong verification name (after reaching confirmationString maxlength).
- Delete `ComponentActions.js`



## After

![zacgwptnvf](https://user-images.githubusercontent.com/4511563/35295084-362c1f30-0046-11e8-99f0-a66199884bce.gif)

<img width="584" alt="screen shot 2018-02-14 at 7 29 21 am" src="https://user-images.githubusercontent.com/4511563/36204656-1cf2f352-115a-11e8-98d2-4a60d5ce20c6.png">

<img width="585" alt="screen shot 2018-02-14 at 6 52 37 am" src="https://user-images.githubusercontent.com/4511563/36204657-1d14084e-115a-11e8-8d8e-1e81769bc0b5.png">




## Ticket

[OSF-8914](https://openscience.atlassian.net/browse/OSF-8914)

pattisdr commented on website/templates/project/nodes_delete.mako:
```diff
@@ -84,6 +85,7 @@
                     </span>
```
The Delete button here can be clicked before the confirmation word is typed.  My opinion, button should be disabled until strings match.
<img width="587" alt="screen shot 2018-01-17 at 2 12 42 pm" src="https://user-images.githubusercontent.com/9755598/35064767-bc3c7eea-fb90-11e7-96c4-5d96e6d84037.png">

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -248,7 +279,7 @@ NodesDeleteViewModel.prototype.confirmChanges =  function() {
+            batchNodesDelete(nodesChanged.reverse(), self.nodeType).then(function () {
                 self.page(self.WARNING);
             }).fail(function (xhr) {
                 $osf.unblock();
```
[L290](https://github.com/CenterForOpenScience/osf.io/pull/7911/files#diff-de94aefe74754843ec4d7c25c5b9275bR290) below, Growl message says "Problem deleting project" regardless of whether this is a project or component.

fabmiz commented:
@pattisdr thanks for the thorough code review. I am done with the changes.Might need product demo input regarding disabling `delete` button till the user inputs the exact confirmation string, but I went ahead and implemented it in all other places.

pattisdr commented:
I'll try to get to this today @fabmiz!

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -145,34 +145,31 @@ var NodesDeleteViewModel = function(node) {
+        }
+
+        return {
+            select: message + (self.isPreprint ? preprintMessage : ''),
```
I know we have a lot of messages here, so maybe this is a product question, but in the QuickDelete case, this is one of the messages, "It will no longer be available to other contributors on the project." Is this also needed in the `NodesDeleteViewModel`.

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -280,24 +282,105 @@ NodesDeleteViewModel.prototype.back = function() {
+    self.nodesDeleted = ko.observable(false);
+};
+
+QuickDeleteViewModel.prototype.handleEditableUpdate = function(element) {
```
Is there any way to dry up some of this duplication? The QuickDeleteViewModel and NodesDeleteModel have a lot of identical sections. This, for example, is identical to `NodesDeleteViewModel.prototype.handleEditableUpdate`

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -280,24 +282,105 @@ NodesDeleteViewModel.prototype.back = function() {
+    self.nodeApiUrl = nodeApiUrl;
+    self.isPreprint = isPreprint;
+    self.message = ko.computed(function(){
+        var preprintMessage = '<br><br>This ' + self.nodeType + ' contains a <strong>preprint</strong>. Deleting this ' +
```
Can we consolidate some of the messages that are identical on the `QuickDeleteViewModel` and `NodesDeleteViewModel`?

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -280,24 +282,105 @@ NodesDeleteViewModel.prototype.back = function() {
-    self.viewModel = new NodesDeleteViewModel(node);
-    self.viewModel.fetchNodeTree().done(function(response) {
-        new NodesDeleteTreebeard('nodesDeleteTreebeard', response, self.viewModel.nodesState, self.viewModel.nodesOriginal);
+QuickDeleteViewModel.prototype.clear = function(){
```
Space between (){}... function() {}.  This happens in a few other places.

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -280,24 +282,105 @@ NodesDeleteViewModel.prototype.back = function() {
+    self.canDelete(inputText === self.confirmationString);
+};
+
+QuickDeleteViewModel.prototype.confirmChanges = function(){
```
space...function() {}

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -245,28 +252,23 @@ NodesDeleteViewModel.prototype.confirmChanges =  function() {
+            if (xhr.responseJSON && xhr.responseJSON.errors) {
+                errorMessage = xhr.responseJSON.errors[0].detail;
+            }
+            $osf.growl('Problem deleting project', errorMessage);
```
Growl message says problem deleting project, regardless of whether this is a project or a component.

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -245,28 +252,23 @@ NodesDeleteViewModel.prototype.confirmChanges =  function() {
-    else {
-        $osf.growl('Verification failed', 'Strings did not match');
+    if (nodesChanged.length <= 100) {
+        $osf.block('Deleting Project');
```
Says deleting project regardless of whether this is a project/component.

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -245,28 +252,23 @@ NodesDeleteViewModel.prototype.confirmChanges =  function() {
-    }
-    else {
-        $osf.growl('Verification failed', 'Strings did not match');
+    if (nodesChanged.length <= 100) {
```
I know that this nodesChanged.length <=100 is a carryover from before, but what happens if it is greater?

fabmiz replied:
@pattisdr, sending this ticket for product demo tomorrow, to make sure we are on the same line regarding recent UI changes before doing anymore add. dev (such addressing this case).

pattisdr replied:
sounds good @fabmiz, thanks for your changes so far!

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -280,24 +282,105 @@ NodesDeleteViewModel.prototype.back = function() {
+        });
+    } else {
+        self.viewModel = new QuickDeleteViewModel(nodeType, isPreprint, nodeApiUrl);
+    }
```
Logic and naming of view models is very readable here: if a project has components, use the NodesDeleteViewModel, otherwise QuickDeleteViewModel. 👍

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -127,68 +117,48 @@ var NodesDeleteViewModel = function(node) {
-                confirm: 'The following components will be deleted.'
-            }[self.page()];
+        return {
+            select: message + (self.isPreprint ? self.preprintMessage : ''),
```
This preprintMessage is only thrown if the parent node is a preprint, but what if you are bulk deleting nodes and a child node is a preprint?  I don't know a quick way to recursively check all the children and make sure there are no preprints, but I think the user needs a warning that they may be inadvertently deleting a preprint.

fabmiz commented on website/static/js/nodesDelete.js:
```diff
@@ -207,97 +177,190 @@ NodesDeleteViewModel.prototype.fetchNodeTree = function() {
+          });
+    });
+
+    return $.when(requests).done(function (_) {
```
Great catch!!

fabmiz commented:
@pattisdr, I finished addressing issues from CR. See the latest attachments. I will have this scheduled from PD and see if they approve the new UI changes.

pattisdr commented on website/static/js/nodesDelete.js:
```diff
@@ -1,114 +1,138 @@
+      ' will also delete your <strong>preprint</strong>. This action is irreversible.';
+
+    self.message = ko.computed(function () {
+        return 'It will no longer be available to other contributors on the project.' +
```
Does this portion of the message need to be in the NodesDeleteViewModel as well?

fabmiz commented on website/templates/project/nodes_delete.mako:
```diff
@@ -19,6 +20,7 @@
                         <div>
                             Select:&nbsp;
                             <a class="text-bigger" data-bind="click:selectAll">All components</a>
+                            <span class="pull-right"><b><sup>*</sup></b><small>project or component contains a preprint</small></span>
```
@pattisdr Product approved of this in the last product demo. If there is no * on any of the listed components, the user knows there is no preprint.

pattisdr replied:
okay, I'll defer to Product, but I found this confusing.

mfraezz commented:
Merged via d3c9f320ed0966cce6af6f9bc963ccf8b5c297e6, closing
